Initial database

JacobHorwath opened the pull request:
Added database create scripts. Tested in SQLFiddle. This version includes contacts and locations entities, which should be enough to have complete base application functionality.

PaulLNH commented on DatabaseScripts/CreateTables.sql:
```diff
@@ -0,0 +1,129 @@
+   entity_address2 TEXT NULL,
+   entity_primary_phone TEXT NULL,
+   entity_secondary_phone TEXT NULL,
+   entity_is_donor bit NULL,
+   entity_is_supplier bit NULL,
```
should there be an `entity_is_healthcare bit NULL,`?

JacobHorwath replied:
Might make sense to have another table for the entity type and then you can have multiple records. Saves having to update the database to add a new type column. This is kind of the quick and dirty tho, it really depends on how the column would be used. What would is_healthcare represent? An entity in the medical line?

JacobHorwath replied:
Yeah I was thinking we needed a distinction between people who donate and people who manufacture, kind of like a 1 time thing vs repeated supplier. But that probably needs rethought. I think a second entity might be useful, but not necessary at this point? We do probably want to have at least 2 bit fields on the entity, one for "customer" and one for "vendors"? Not sure what to call exactly, suppliers, demanders?

PaulLNH commented on DatabaseScripts/CreateTables.sql:
```diff
@@ -0,0 +1,129 @@
+   entity_business_name TEXT NOT NULL,
+   entity_first_name TEXT NULL,
+   entity_last_name TEXT NULL,
+   entity_address1 TEXT NULL,
+   entity_address2 TEXT NULL,
```
Should this just be one filed linking to a row in the `locations` table?

JacobHorwath replied:
More of an old habit for mailing use: Address line 1, Address line 2, you know how you would put your address on line one in the mail and then the apartment number on line 2

PaulLNH commented on DatabaseScripts/CreateTables.sql:
```diff
@@ -0,0 +1,129 @@
+    location_zip TEXT NOT NULL,
+    location_state TEXT NOT NULL,
+    location_country TEXT NOT NULL,
+    location_primary_phone TEXT NOT NULL,
```
This could probably be null, or not in here at all. Not all "locations" have a phone, the contact details should probably be at the "contacts" or "entities" level

JacobHorwath replied:
How does one ship to a location without an address? Maybe phone shouldn't be required, but I think all things that you would need to ship

PaulLNH replied:
I agree, this comment was for `location_primary_phone` only. Locations should be a physical place on a map and a location would be tied to an entity (organization, healthcare, or donor) and that entity would have contact details. Let me know if I'm thinking of this incorrectly.

JacobHorwath replied:
Yeah, I was thinking a warehouse would need a land phone line, but we can remove it and the user could always put a phone number in the notes field if they need to

PaulLNH commented on DatabaseScripts/CreateTables.sql:
```diff
@@ -0,0 +1,129 @@
+   web_id uuid DEFAULT uuid_generate_v4 (),
+   entity_id integer REFERENCES entities(entity_id) NOT NULL,
+   location_id integer REFERENCES locations(location_id) NULL,
+   contact_business_name TEXT NULL,
```
I think I like this field but I want to make sure we are on the same page with it's intended usage. How is this different from an `entities_business_name`? If these are linked would this be redundant or is there a scenario where these can be unique?

JacobHorwath replied:
I thought there might be a scenario where it could be unique, like a contact for a company that doesn't actually work at that company (contractor?)

PaulLNH commented on DatabaseScripts/CreateTables.sql:
```diff
@@ -0,0 +1,129 @@
+);
+
+
+CREATE TABLE items(
+	item_id serial PRIMARY KEY NOT NULL,
+    web_id uuid DEFAULT uuid_generate_v4 (),
+	item_number TEXT NULL,
+	item_description TEXT NULL,
+	item_upc TEXT NULL,
+    created_on TIMESTAMP NOT NULL DEFAULT CURRENT_DATE
+);
```
Are items "finished products"? Say we have 3 sheets of fabric and can make 6 gowns and 12 masks out of it, would each one of those be an item? I'm assuming there is no BoM with this setup and that's fine, just trying to determine how we would add a value to this. It would be nice to say "We've helped NH Healthcare facilities receive $100,000 worth of supplies" but that may be a moot point, "We've helped NH Healthcare facilities get 50,000 personal protective products for their staff and patients." would probably be just as good.

JacobHorwath replied:
It's not currently built for bill of materials or manufacturing purpose. That's kind of a big rabbit hole, but we can definitely go there if needed. Lots of recursion and complexity. This is the master item list (like the product catalog equivalent). Reporting on transactions would be done on the fulfillment table.

JacobHorwath replied:
So yes, items are finished products. But we're not building the BOMs here. If someone needed parts for their finished goods they would just put in a request for the components too, currently.

PaulLNH commented on DatabaseScripts/CreateTables.sql:
```diff
@@ -0,0 +1,129 @@
+    created_on TIMESTAMP NOT NULL DEFAULT CURRENT_DATE
+);
+
+CREATE TABLE orders(
+	order_id serial PRIMARY KEY NOT NULL,
+    web_id uuid DEFAULT uuid_generate_v4 (),
+	entity_id integer REFERENCES entities(entity_id) NOT NULL,
+    request_id integer REFERENCES requests(request_id) NOT NULL,
+	order_description TEXT NULL,
+	order_created_date TIMESTAMP NOT NULL DEFAULT CURRENT_DATE,
+	order_due_date TIMESTAMP NOT NULL,
+    created_on TIMESTAMP NOT NULL DEFAULT CURRENT_DATE
+);
```
How are orders different from requests? I'm just trying to visualize the workflow here. A request is something a facility needs, an order is an organization or group providing materials and fufillment is the actual transaction?

JacobHorwath replied:
Yes, Requests is something a facility needs. So the facility would set up a request with a list of items. Or multiple requests (for different locations perhaps) with multiple items on each. Orders are the supplier side equivalent. Then each order could have multiple fulfillments (shipments). Say if they could partially fulfill an order.

JacobHorwath commented:
Ok I reviewed the spec again and added a few boolean columns. Added itemtypes table so that we can differentiate between materials, money, facilities and time. I also made changes based on the previous discussion.
